### PR TITLE
feat: add power of two scaling adapter for roundPBS

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
- Adds a pattern matcher for detecting the Power of Two scaling usage:
- Fixes a bug in the onnx parser when an initalizer node is used by several nodes
- Fixes a bug in the LLM notebook for reshape semantics
- add tests with custom networks with compile_brevitas_qat
- add tests when rounding is set manually and should not be overwritten
- add tests for correctness -> not working well yet: https://github.com/zama-ai/concrete-ml-internal/issues/3946

Closes https://github.com/zama-ai/concrete-ml-internal/issues/3947
Closes https://github.com/zama-ai/concrete-ml-internal/issues/3946

### CODE REVIEW

File: use_case_examples/llm/utils.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def max_fhe_relu(q_x, axis=-1, keepdims=True):
         if keepdims:
             shape = list(result.shape)
             shape.insert(axis, 1)
-            result = result.reshape(shape)
+            result = result.reshape(tuple(shape))
```
Comment (andrei-stoian-zama) [inline use_case_examples/llm/utils.py:35]: CP changed the semantics, they only accept tuples now

File: src/concrete/ml/onnx/ops_impl.py [hunk 2]
```diff
@@ -1140,7 +1139,6 @@ def numpy_transpose(x: numpy.ndarray, *, perm=None) -> Tuple[numpy.ndarray]:
     return (numpy.transpose(x, axes=perm),)
 
 
-@onnx_func_raw_args("b")
```
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/ops_impl.py:1143]: same

File: src/concrete/ml/pytest/torch_models.py [hunk 3]
```diff
@@ -781,47 +823,6 @@ def forward(self, x):
         x = self.fc1(x)
         return x
 
-    def test_torch(self, test_loader):
```
Comment (RomanBredehoft) [inline src/concrete/ml/pytest/torch_models.py:784]: why was this removed ?

File: src/concrete/ml/quantization/post_training.py [hunk 4]
```diff
@@ -604,6 +607,9 @@ def quantize_module(self, *calibration_data: numpy.ndarray) -> QuantizedModule:
             onnx_model=self.numpy_model.onnx_model,
         )
 
+        adapter = PowerOfTwoScalingRoundPBSAdapter(quantized_module)
+        adapter.process()
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/post_training.py:611]: what is "process" ?

File: src/concrete/ml/quantization/quantized_module_passes.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+        Args:
+            predecessors (PredecessorsType): Module predecessor operation list
+            nodes_in_path (List[QuantizedOp]): list of quantized ops in the pattern
+            input_producer_of_path (Optional[QuantizedOp]): operation that produces the input
+
+        Returns:
+            result (bool): whether the pattern can be optimized
+        """
+
+        if len(nodes_in_path) != 3:
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module_passes.py:135]: all of this method's steps need at least a comment explaining why we return these values as it's not very clear

File: src/concrete/ml/pytest/torch_models.py (hunk 3 above)
Comment (andrei-stoian-zama) [reply src/concrete/ml/pytest/torch_models.py:784]: it was converted to a simple function instead of a class method. Thus it can be used on other model instances

File: src/concrete/ml/quantization/base_quantized_op.py [hunk 6]
```diff
@@ -962,7 +962,7 @@ def cnp_round(
                 self.lsbs_to_remove = current_n_bits_accumulator - self.rounding_threshold_bits
 
             # mypy
-            assert self.lsbs_to_remove is not None
+            assert self.lsbs_to_remove is not None and self.lsbs_to_remove >= 0
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/quantization/base_quantized_op.py:965]: good catch, no, It's not a mypy issue

File: src/concrete/ml/quantization/qat_quantizers.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    MaxStatsScaling,
+    ParamFromRuntimePercentileScaling,
+    PerTensorPoTScaling8bit,
+    WeightQuantSolver,
+)
+from brevitas.quant.solver.act import ActQuantSolver
+
+
+# pylint: disable-next=too-many-ancestors
+class Int8ActPerTensorPoT(
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/quantization/qat_quantizers.py:13]: this needed a separate file because otherwise lazydocs failed when parsing these classes that inherit brevitas, so I needed to ignore only these classes in lazydocs .

File: src/concrete/ml/quantization/quantized_module_passes.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+            Args:
+                value (float): the value for which to take the log2
+
+            Returns:
+                result: The integer log2 and a bool indicating whether
+                    the input value was an integer power of two
+            """
+            log2_value = int(numpy.rint(numpy.log2(value)))
+            if numpy.isclose(numpy.power(2.0, log2_value), value, atol=0.01):
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module_passes.py:239]: is it not enough t just check that `numpy.rint(numpy.log2(value)) == numpy.log2(value)` ? or am I missing something ? 

if not, then maybe make "atol" a parameter for `integer_log2` (or at least say why it was set to 0.01)

Comment (andrei-stoian-zama) [reply src/concrete/ml/quantization/quantized_module_passes.py:239]: the best practice is to never compare floats using equality. But you're right about the 0.01, it's quite arbitrary and not a good value for low powers of two (it's too big when the power is say -7). I'll use rtol instead

File: src/concrete/ml/quantization/quantized_ops.py [hunk 9]
```diff
@@ -749,6 +765,16 @@ def q_impl(
...
 
+        if (
+            q_bias is not None
+            and isinstance(q_bias, QuantizedArray)
+            and q_bias.quantizer.is_precomputed_qat
+        ):
+            # Make sure the scale was correctly matching during training
+            # The bias scale should be the same scale as the one of the weights * inputs
+            assert numpy.isclose(q_bias.quantizer.scale, m_matmul)
+            numpy_q_out += q_bias.qvalues.reshape((1, -1, 1, 1))
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_ops.py:776]: why the reshape this time ?

File: tests/torch/test_brevitas_qat.py [hunk 10]
```diff
@@ -405,3 +492,102 @@ def test_brevitas_constant_folding(default_configuration):
...
+            len(round_pbs_patterns) == 0
+        ), "Expected number of round PBS optimized patterns was not matched"
+        # 3 layers
+        assert (
+            adapter.num_ignored_valid_patterns == 3 - 1
+        ), "Expected number of ignored round PBS optimizable patterns was not matched"
+    else:
+        pass
+
+    # y_pred_clear_round = model.predict(x_test, fhe="disable")
```
Comment (RomanBredehoft) [inline tests/torch/test_brevitas_qat.py:558]: is the following supposed to be removed ?

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
huge work, thanks a lot ! I have several comments (mostly things about detailing a bit the steps with comments)

Also, is it expected that apidocs were generated in this PR ? we'll update them before releasing anyway !

File: src/concrete/ml/quantization/qat_quantizers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (RomanBredehoft) [reply src/concrete/ml/quantization/qat_quantizers.py:13]: maybe add it as a comment somewhere then, it's not very obvious at first glance !

File: src/concrete/ml/onnx/ops_impl.py [hunk 11]
```diff
@@ -240,7 +240,6 @@ def numpy_constant(**kwargs):
 # pylint: disable=invalid-name
 # 1 is technically an int but is accepted by mypy as a float (and it simplifies our life for
 # compilation) so instead of passing 1.0 by default 1 is passed
-@onnx_func_raw_args("c")
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/onnx/ops_impl.py:243]: so that it can be a either a quantized value or a raw value. Setting it to "raw" means it's always a raw float value

File: src/concrete/ml/quantization/base_quantized_op.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/quantization/base_quantized_op.py]: removed it, it wasn't doing anything anyway

File: src/concrete/ml/onnx/ops_impl.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/ops_impl.py:243]: ok thanks for explaining !

File: src/concrete/ml/quantization/post_training.py [hunk 12]
```diff
@@ -427,13 +428,13 @@ def _quantize_layers(self, *input_calibration_data: numpy.ndarray):
                 attributes.update({"rounding_threshold_bits": self.rounding_threshold_bits})
 
             # All inputs, allow optional constants (they become None)
-            curr_inputs = {
-                input_name: node_results.get(input_name, None) for input_name in node.input
-            }
+            curr_inputs = [
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/quantization/post_training.py:432]: If you could pin-point what is not clear it would help.. I don't really know which parts to explain

File: tests/torch/test_brevitas_qat.py (hunk 10 above)
Comment (andrei-stoian-zama) [reply tests/torch/test_brevitas_qat.py]: I added it back but correctness is not achieved, I'd rather push it as is and work on it later. I created an issue on it

File: src/concrete/ml/quantization/post_training.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (RomanBredehoft) [reply src/concrete/ml/quantization/post_training.py:432]: yes sure !

File: src/concrete/ml/quantization/post_training.py [hunk 13]
```diff
@@ -427,13 +428,13 @@ def _quantize_layers(self, *input_calibration_data: numpy.ndarray):
...
+                (input_name, node_results.get(input_name, None)) for input_name in node.input
+            ]
 
             # Constant inputs
             curr_cst_inputs: Dict[int, ONNXOpInputOutputType] = {}
-            for input_idx, (input_name, value) in enumerate(curr_inputs.items()):
+            for input_idx, (input_name, value) in enumerate(curr_inputs):
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/post_training.py:438]: what does this section do overall ?

File: src/concrete/ml/quantization/post_training.py [hunk 14]
```diff
@@ -455,10 +456,12 @@ def _quantize_layers(self, *input_calibration_data: numpy.ndarray):
...
+                input_name for input_name, _ in curr_inputs if input_name not in constants
             ]
             curr_calibration_data = tuple(
-                curr_inputs[input_name] for input_name in variable_input_names
+                input_data
+                for input_name, input_data in curr_inputs
+                if input_name in variable_input_names
             )
 
             # For mypy
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/post_training.py:468]: below this ( I can't comment further here) : 
- why casting is need : ` curr_calibration_data = cast(Tuple[numpy.ndarray], curr_calibration_data)`
- "# Find the unique integer producers of the current's op output tensor" what is a producer ? how are they used ?
-

File: src/concrete/ml/quantization/post_training.py [hunk 15]
```diff
@@ -455,10 +456,12 @@ def _quantize_layers(self, *input_calibration_data: numpy.ndarray):
             has_variable_inputs = (len(curr_inputs) - len(curr_cst_inputs)) > 0
 
             variable_input_names = [
-                input_name for input_name in curr_inputs if input_name not in constants
+                input_name for input_name, _ in curr_inputs if input_name not in constants
             ]
             curr_calibration_data = tuple(
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/post_training.py:462]: basically we are only interested in data from variables right ?

**Comment by fd0r** [comment]
Conformance issue.

**Review by fd0r — APPROVED** [review]

**Review by RomanBredehoft — DISMISSED** [review]
unblocking my request for changes, fine for me if you take care of the remaining comments in a following PR !

**Comment by jfrery** [comment]
Should we also update a notebook to make sure there is a speed up? Would be great to add a definition about this quantizer. I.e. what it does and what speed up is expected. I suppose the gain would come from using rounding instead of a standard PBS to re-quantize?

**Comment by fd0r** [comment]
That's a very long CI, unsure what to do here 🤔

File: docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb [hunk 16]
```diff
@@ -62,16 +62,13 @@
...
    "metadata": {},
    "outputs": [],
    "source": [
     "params = {\n",
     "    \"module__n_layers\": 3,\n",
-    "    \"module__n_w_bits\": 3,\n",
-    "    \"module__n_a_bits\": 4,\n",
-    "    \"module__n_accum_bits\": 9,\n",
-    "    \"module__activation_function\": nn.Sigmoid,\n",
+    "    \"module__activation_function\": nn.ReLU,\n",
```
Comment (jfrery) [inline docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb:71]: Why was this needed? Changing the architecture might have a positive impact on the accuracy which is not really what we want in this PR.

File: docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb [hunk 17]
```diff
@@ -198,21 +195,21 @@
...
-      "100%|██████████| 38/38 [02:38<00:00,  4.16s/it]"
+      "100%|██████████| 38/38 [00:46<00:00,  1.23s/it]"
      ]
     },
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "Execution time: 4.17 seconds per sample\n"
+      "Execution time: 1.23 seconds per sample\n"
```
Comment (jfrery) [inline docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb:212]: Nice!

File: use_case_examples/llm/utility_functions.py [hunk 18]
```diff
@@ -32,7 +32,7 @@ def max_fhe_relu(q_x, axis=-1, keepdims=True):
         if keepdims:
             shape = list(result.shape)
             shape.insert(axis, 1)
-            result = result.reshape(shape)
+            result = result.reshape(tuple(shape))
```
Comment (RomanBredehoft) [inline use_case_examples/llm/utility_functions.py:35]: is this change expected ? 🤔

**Review by RomanBredehoft — APPROVED** [review]
looks great, thanks a lot for this feature !!

File: docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb (hunk 16 above)
Comment (RomanBredehoft) [reply docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb:71]: I believe this is because the power of two scaling feature only works with ReLU

File: src/concrete/ml/sklearn/qnn_module.py [hunk 19]
```diff
@@ -27,13 +28,15 @@ def __init__(
         n_layers: int,
         n_outputs: int,
         n_hidden_neurons_multiplier: int = 4,
-        n_w_bits: int = 3,
-        n_a_bits: int = 3,
-        n_accum_bits: int = MAX_BITWIDTH_BACKWARD_COMPATIBLE,
+        n_w_bits: int = 4,
+        n_a_bits: int = 4,
+        # No pruning by default as roundPBS keeps the PBS precision low
```
Comment (jfrery) [inline src/concrete/ml/sklearn/qnn_module.py:33]: What' is the chosen bit width to round down?

File: use_case_examples/llm/utility_functions.py (hunk 18 above)
Comment (andrei-stoian-zama) [reply use_case_examples/llm/utility_functions.py:35]: it's just an additional fix that could have used a separate PR but got stuck in this one

File: src/concrete/ml/sklearn/qnn_module.py (hunk 19 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/qnn_module.py:33]: So these hyper parameters are useless? I am trying to understand what is the number of bits that's used to round?

File: use_case_examples/llm/utility_functions.py (hunk 18 above)
Comment (RomanBredehoft) [reply use_case_examples/llm/utility_functions.py:35]: got it thanks

File: docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb (hunk 16 above)
Comment (jfrery) [reply docs/advanced_examples/FullyConnectedNeuralNetwork.ipynb:71]: Got it. But I don't see anything that prevents the use of sigmoid or other non linear function. What would happen then?

File: src/concrete/ml/sklearn/qnn_module.py [hunk 20]
```diff
@@ -27,13 +28,15 @@ def __init__(
...
+        n_accum_bits: int = 32,
         n_prune_neurons_percentage: float = 0.0,
         activation_function: Type = nn.ReLU,
         quant_narrow: bool = False,
         quant_signed: bool = True,
+        power_of_two_scaling: bool = True,  # Default to true: use roundPBS to speed up the NNs
```
Comment (jfrery) [inline src/concrete/ml/sklearn/qnn_module.py:39]: It's a bit weird here how rounding is implicitly contained in the PoT feature. Shouldn't there be 2 distinct features?

**Review by fd0r — APPROVED** [review]
Approved but some experiments/documentation should be added to explain to the user why and how this works.
